### ISSUE TITLE
UnnecessaryParentheses false positives in 8.42

### ISSUE THREAD

**Issue by inponomarev** [checkstyle/checkstyle#9957]
This must be related to https://github.com/checkstyle/checkstyle/issues/5779 -- this behavior appeared in 8.42:

```bash
$ cat TestClass.java
public class TestClass {
    public static void main(String[] args) {
        boolean a = true;
        int b = 42;
        int c = 42;
        System.out.println(a == (b == c));
    }
}
```

```bash
$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
         <module name="UnnecessaryParentheses" />
    </module>
</module>
```

```bash
$ java -jar checkstyle-8.42-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] /mnt/c/Users/ipono/Downloads/__files/TestClass.java:6:33: Unnecessary parentheses around expression. [UnnecessaryParentheses]
Audit done.
Checkstyle ends with 1 errors.
```

**Comment by strkkk** [comment]
@Vyom-Yadav please take a look

**Comment by Vyom-Yadav** [comment]
I am on it.

**Comment by rnveach** [comment]
It seems to me this is only a false positive because of the types of the variables.
````
        boolean a = true;
        int b = 42;
        int c = 42;
````

A compile error occurs with the `()s` removed because we are comparing boolean to integer. The compiler goes from left to right and so it makes sense.

Changing everything to boolean and removing the `()`s, the code compiles fine and I don't see any reason why it isn't valid.
````
		boolean a = true;
		boolean b = false;
		boolean c = true;
		System.out.println(a == b == c);
````

Additionally, changing just the first 2 variables to the same type like integer still compiles and looks valid. Left to right, the first equals gets done first which results a boolean and can be compared to the last variable.
````
		int a = 0;
		int b = 1;
		boolean c = true;
		System.out.println(a == b == c);
````

It seems to me this is strictly only a false positive based on the types of the variables which is leaning extremely closely to our limitation that we can't identify the type of expressions ( https://checkstyle.org/writingchecks.html#Limitations ).

@romani Give your feedback on how we should handle this case.

It seems to me we should suppress anything in the form `a == (b {Comparison Operator} c)` and this would include all operators that result in a boolean result like `==`, `!=`, etc...

**Comment by Vyom-Yadav** [comment]
Agreed @rnveach also currently the check analyses based on the token types i.e. `==`, `!=` ,etc. To get no violation for these types will be almost like evaluating the whole expression. We would have to check the type for the identifier and keep a note of it. Even if we keep the note of type of variable, it would further increase complexity whether it is a local or an instance variable. Also by suppressing these cases we will be dropping many cases which were **really** using extra parentheses.

**Comment by Vyom-Yadav** [comment]
@romani Kindly advise how we should handle this case.

**Comment by romani** [comment]
We should not place violation on such code.

>It seems to me we should suppress anything in the form

Yes.

**Comment by romani** [comment]
@Vyom-Yadav , please do a fix. If fix is complicated, we will need to to revert your changes that cause such problems.

**Comment by dejan2609** [comment]
@romani POC _(apache kafka with CheckStule 8.42)_ is provided; LMK in case that you are missing some details or you want me to change something.

> We need to add Kafka to our regression suit [apache/kafka#10656](https://github.com/apache/kafka/pull/10656) , and make sure there is no false positives.

FYI @ijuma

**Comment by romani** [comment]
fix is merged

**Comment by romani** [comment]
@Vyom-Yadav , thanks a lot for quick fix.
